prov/opx/src/fi_opx_init.c

```diff
@@ -0,0 +1,481 @@
+				"unavailable [bad info->addr_format (%u)]",
+				info->addr_format);
+		goto err;
+	}
+
+	if (info->tx_attr) {
+		ret = fi_opx_check_tx_attr(info->tx_attr);
+		if (ret)
+			return ret;
+	}
+
+	if (info->rx_attr) {
+		ret = fi_opx_check_rx_attr(info->rx_attr);
+		if (ret)
+			return ret;
+	}
+
+	if (info->ep_attr) {
+		ret = fi_opx_check_ep_attr(info->ep_attr);
+		if (ret)
+			return ret;
+	}
+
+	if (info->domain_attr) {
+		ret = fi_opx_check_domain_attr(info->domain_attr);
+		if (ret)
+			return ret;
+	}
+
+	if (info->fabric_attr) {
+		ret = fi_opx_check_fabric_attr(info->fabric_attr);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+
+err:
+
+	errno = FI_ENODATA;
+	return -errno;
+}
+
+static int fi_opx_fillinfo(struct fi_info *fi, const char *node,
+		const char* service, const struct fi_info *hints,
+	        uint64_t flags)
+{
+	int ret;
+	uint64_t caps;
+	union fi_opx_addr *addr, *src_addr, *dest_addr;
+	uint32_t fmt;
+	size_t len;	
+
+	if (!fi)
+		goto err;
+
+	if (!hints && !node && !service)
+		goto err;
+
+	if (hints && (((hints->mode & FI_CONTEXT) != 0) && ((hints->mode & FI_CONTEXT2) == 0))) {
+		FI_WARN(fi_opx_global.prov, FI_LOG_FABRIC,
+			"FI_CONTEXT mode is not supported. Use FI_CONTEXT2 mode instead.\n");
+		errno = FI_ENODATA;
+		return -errno;
+	}
+
+	fi->next = NULL;
+	fi->caps = FI_OPX_DEFAULT_CAPS;
+
+	/* set the mode that we require */
+	fi->mode = FI_ASYNC_IOV;
+	fi->mode |= (FI_CONTEXT2);
+
+	/* clear modes that we do not require */
+	fi->mode &= (~FI_LOCAL_MR);
+	fi->mode &= (~FI_MSG_PREFIX);
+	fi->mode &= (~FI_CONTEXT);
+
+	fi->addr_format = FI_ADDR_OPX;
+	fi->src_addrlen = sizeof(union fi_opx_addr);
+	fi->dest_addrlen = sizeof(union fi_opx_addr);
+
+	src_addr = dest_addr = NULL;
+
+	// Process the node field. Service is treated identically to node.
+	if (node) {
+		if (!ofi_str_toaddr(node, &fmt, (void **)&addr, &len) &&
+	    	fmt == FI_ADDR_OPX) {
+			if (flags & FI_SOURCE) {
+				src_addr = addr;
+				FI_INFO(fi_opx_global.prov, FI_LOG_FABRIC,
+					"'%s' is taken as src_addr.\n", node);
+			} else {
+				dest_addr = addr;
+				FI_INFO(fi_opx_global.prov, FI_LOG_FABRIC,
+					"'%s' is taken as dest_addr.\n", node);
+			}
+			node = NULL;
+		} else {
+			FI_WARN(fi_opx_global.prov, FI_LOG_FABRIC,
+				"'%s' is not a valid OPX address.\n", node);
+			goto err;
+		}
+	}
+
+	if (hints) {
+		if (hints->dest_addr) {
+			FI_LOG(fi_opx_global.prov, FI_LOG_DEBUG, FI_LOG_FABRIC,
+				"cannot support dest_addr lookups now\n");
+			goto err;
+		} else if (hints->src_addr) {
+			src_addr = mem_dup(hints->src_addr, hints->src_addrlen);
+		} else {
+			// MHEINZ: WTH is this supposed to accomplish?
+			src_addr = (union fi_opx_addr*)((service)?strdup(service):NULL);
+		}
+
+	}
+
+	if (!src_addr) {
+		src_addr = calloc(1, sizeof(*src_addr));
+		if (!src_addr) {
+			FI_WARN(fi_opx_global.prov, FI_LOG_FABRIC, "Failed to alloc memory.\n");
+			goto err;
+		}
+		src_addr = mem_dup(&opx_default_addr, sizeof(opx_default_addr));
```

Comment:
Either mem_dup or calloc/copy, but not both.  The calloc region was just leaked.

Reply:
I will address this in the next push.